### PR TITLE
Add Mathjax support.

### PR DESCRIPTION
**Pull request by adamwolf**
## Purpose / Description
Add Mathjax support.

## Fixes
https://github.com/ankidroid/Anki-Android/issues/4915

## Approach
I modeled the approach after desktop Anki 2.1,  omitting the fade in order to save every last millisecond. :)

Nearly everything outside of tests is a translation of things that are in Anki 2.1, and the tests are mostly cribbed as well.

Let me know if you don't like the submodule.  If so, feel free to just commit the files directly, or have me do it.

## How Has This Been Tested?

I added some automated cloze tests from Anki.  I added the Mathjax ones as well as non-Mathjax ones.

I used the Android emulator, and setup cards with text of \( x_i \) and \[\ce{C6H5-CHO}\] to test both Mathjax and mhchem.

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x]  You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code

### CODE REVIEW

**Comment by mikehardy** [comment]
Commentary on issues is fast and easy but looking at code takes me a while as I have to re-read it a few times. Give me some time but I will go through this starting now as time allows

**Comment by adamwolf** [comment]
Thank you.  No problem, and no rush.

I understand the dedication and commitment it takes to maintain a project,
and I applaud and thank you (and everyone else) for your efforts.

If folks are looking to get mhchem working with the current workaround with
JavaScript in the cards, they could use the mods in this to work out how to
get that going.

**Comment by timrae** [comment]
Awesome, thank you for doing this!

> Let me know if you don't like the submodule. If so, feel free to just commit the files directly, or have me do it.

Yeah please remove the submodule, we don't use them. Dependencies should either be handled in the build.gradle file or committed directly into the source tree, which ever you prefer.

File: AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java [hunk 1]
```diff
@@ -285,29 +285,91 @@ public int compare(String lhs, String rhs) {
     }
 
     private static String clozeText(String txt, String ord, char type) {
-        Matcher m = Pattern.compile(String.format(Locale.US, clozeReg, ord)).matcher(txt);
-        if (!m.find()) {
+        Matcher n = Pattern.compile(String.format(Locale.US, clozeReg, ord)).matcher(txt);
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java]: This variable ` Matcher n` is only used once right? I think readability would be improved by making this a one-liner so we don't have both `m` and `n` variables.

File: AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java [hunk 2]
```diff
@@ -285,29 +285,91 @@ public int compare(String lhs, String rhs) {
...
+                    buf = "[...]";
                 }
             } else {
-                m.appendReplacement(repl, "<span class=cloze>$1</span>");
+                buf = m.group(2);
             }
+            StringBuilder replacement = new StringBuilder();
+
+            if (m.group(1).equals("c")) {
+                replacement.append("<span class=cloze>");
```
Comment (timrae) [inline AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java]: Can you use `String.format()` to make it closer to the Python?

**Comment by adamwolf** [comment]
It may be a few days until I can get to these changes.  Thanks for the fast
response!

**Comment by adamwolf** [comment]
> Yeah please remove the submodule, we don't use them. Dependencies should either be handled in the build.gradle file or committed directly into the source tree, which ever you prefer.

This will be fixed shortly.

File: AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java (hunk 1 above)
Comment (adamwolf) [reply AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java]: Done.

File: AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java (hunk 2 above)
Comment (adamwolf) [reply AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java]: Done.

**Comment by mikehardy** [comment]
Just a note that my previous "fix" for CI that enabled the X86 emulators now seems to have paradoxically blown up the older APIs with arm emulaturs. So CI is failing but if you inspect it is a clean split between arm failing and x86 working. I'm not happy about that of course but I wouldn't consider it a blocker so long as it remains in that state

**Comment by adamwolf** [comment]
I will be away on travel for the next few days, so please do not assume I have abandoned this PR if I do not respond promptly :)

File: AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java (hunk 1 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java]: looks good

**Comment by mikehardy** [comment]
I ran this locally today and I was able to see MathJax in cards! :beer: 

However, the test collection from the linked issue: https://github.com/ankidroid/Anki-Android/issues/4915#issuecomment-433043958 does not run - the difference is the cloze card type - there is something going on with cloze/mathjax combo processing that isn't rendering the MathJax. I noticed in the tests when I executed `./gradlew jacocoTestReport` with an emulator running, then looked at `./AnkiDroid/build/reports/jacoco/jacocoTestReport/html/index.html` that the cloze-related stuff was the only stuff without coverage

Putting these together, I'm guessing if you extend your current unit test to cover the cloze cases in the provided collection hopefully be something obvious in a regular expression or similar

**Comment by adamwolf** [comment]
Sounds good.  I am on travel but I will get to this in the next week or two.

**Comment by adamwolf** [comment]
I dug into this a bit today.  I think I get most of what's going on, but I don't know the right choice.

The cards in this deck use a workaround and do not use the native Anki Desktop Mathjax support.  If you remove the extra scripts from the templates imported when you import this deck, the cards appear broken in Anki Desktop too.

I do not understand why the changes I've made here have made the template-mathjax-workaround *not* work on AnkiDroid but the same deck does work on Anki Desktop.  I will investigate, but if anyone else has time to take a look it would be appreciated too.

**Comment by mikehardy** [comment]
Hey @adamwolf - curious if you saw anything in your investigation? If I understand correctly, "pure Mathjax" cards (those not using the stop-gap workaround) work fine, it's just ones that are using the workaround that do not work?

**Comment by adamwolf** [comment]
That matches my observations (and the existing tests, copied from Anki).  I
haven't yet been able to dig into this but I haven't abandoned it.

**Comment by adamwolf** [comment]
Please note the ones with the workaround work fine with Desktop Anki so
there isn't a fundamental issue supporting both.

**Comment by mikehardy** [comment]
I just built this as a parallel release for possible wider testing - people may find it here: https://github.com/mikehardy/Anki-Android/releases/tag/2.9alpha68-mathjax

But I saw that inclusion of Mathjax adds 37MB to the size of the APK (it goes from 10MB to 47MB!). It makes sense in a way, but this could be an issue we need to handle, that's a very sizable jump. 

@timrae how would you feel about trying the new fancy APK splitting stuff where optional features are only downloaded for people that need them? Then we could hive this stuff off, while still making a universal build for non-Play store people? I personally hate the idea of complicating the build but I'm not sure how to have Mathjax and not have it be huge...

Screen Shot 2019-06-03 at 12 38 45 AM

**Comment by mikehardy** [comment]
@adamwolf I just force-pushed on your branch - only changes were to rebase you to upstream/master and add a gradle.settings entry that was required for me to build the demo I reference above. No functional changes at all but you will need to `git reset --hard` in your checkout to get up to date

**Comment by mikehardy** [comment]
This is what Mathjax recommends for cleaning - we need to make some decisions on what fonts we will support etc and do post-processing on the distribution.

https://github.com/mathjax/MathJax-docs/wiki/Guide%3A-reducing-size-of-a-mathjax-installation
which points to 
https://github.com/mathjax/MathJax-grunt-cleaner/blob/master/Gruntfile.js

Based on the [revision history of the file](https://github.com/mathjax/MathJax-docs/wiki/Guide:-reducing-size-of-a-mathjax-installation/1814429ed1e97bfb7675c0fd400804baa9287249) it appears it can be as small as 700Kb if we restrict the input/output severely so this should be the first size optimization path

We need to decide precisely which input formats, output formats, and fonts to support and that will drive the APK size increase. This work may have already been done in Anki Desktop, we should in general support the input formats they support I think for ecosystem compatibility, but we can make Android-specific output/font decisions.

**Comment by adamwolf** [comment]
Thanks Mike!  I've been busy changing jobs but this is on my
"important projects" list too...

I did not look at what has already been done on this front from Anki
Desktop.  That's a good idea.  (No wonder it took so long to transfer
to my phone...)

**Comment by mikehardy** [comment]
Now that I know how big the APK is without shrinking, I can understand why CI is failing. On the 3 ARM emulators it just times out while copying the APK onto the emulator in Travis. This should resolve once we crunch it down and CI should go green. Good luck with the new job @adamwolf

**Comment by timrae** [comment]
> Based on the revision history of the file it appears it can be as small as 700Kb

Nice find!

> We need to decide precisely which input formats, output formats, and fonts to support and that will drive the APK size increase. This work may have already been done in Anki Desktop, we should in general support the input formats they support I think for ecosystem compatibility, but we can make Android-specific output/font decisions.

Sounds good to me.
@dae any insights to share?

**Comment by dae** [comment]
I'd suggest you just use the web/mathjax folder from Anki - it has already been shrunk down to 1.9M, and that's going to be the most compatible option.

**Comment by adamwolf** [comment]
I like this idea.  I know we'll get folks worrying about exact
compatibility already, (rightfully so!) so as many things that can be
exactly the same is great.

**Comment by mikehardy** [comment]
Of course, thanks @dae

**Comment by mikehardy** [comment]
I just swapped out the previous mathjax for the upstream folder, and it's 723Kb in the APK zipped. Very very nice. I'm doing a parallel release now for testing, but in my testing this works just fine normal MathJAX

@timrae the sole problem is that there is a problem with the stop-gap compatibility template. I want to merge this and move on, just telling the compatibility people (there are probably like 5) to revert their stopgap, go with raw Mathjax now and be done with it.

**Comment by adamwolf** [comment]
I suspect I can figure out the stopgap compatibility thing this week, or at
least know more about it to know that it isn't hiding a deeper issue.

**Comment by timrae** [comment]
Fine, let's release and fix issues later on this one

**Comment by adamwolf** [comment]
👍

**Comment by timrae** [comment]
Thank you for adding this great feature, sayonara LaTeX

**Pull request merged**